crates/common/checkpoint_sync/src/lib.rs

```diff
@@ -0,0 +1,146 @@
+///     "execution_optimistic" : Some("false"),
+///     "finalized" : None,
+///     "data" : T
+/// })
+/// }
+#[derive(Debug, Serialize, Deserialize)]
+pub struct OptionalBeaconVersionedResponse<T> {
+    pub version: Option<String>,
+    pub execution_optimistic: Option<Value>,
+    pub finalized: Option<Value>,
+    pub data: T,
+}
+
+/// Entry point for checkpoint sync.
+pub async fn initialize_db_from_checkpoint(
+    db: ReamDB,
+    checkpoint_sync_url: Option<Url>,
+) -> anyhow::Result<()> {
+    if db.is_initialized() {
+        warn!("Starting Checkpoint Sync from existing DB. It is advised to start from a clean DB");
+        return Ok(());
+    }
+
+    let checkpoint_sync_url = get_checkpoint_sync_sources(checkpoint_sync_url).remove(0);
+    info!("Initiating Checkpoint Sync");
+    let block = fetch_finalized_block(&checkpoint_sync_url).await?;
+    info!(
+        "Downloaded Block: {} with Root: {}. Slot: {}",
+        block.message.body.execution_payload.block_number,
+        block.message.block_root(),
+        block.message.slot
+    );
+    let slot = block.message.slot;
+    initialize_blobs_in_db(&checkpoint_sync_url, db.clone(), block.message.block_root()).await?;
+    info!(
+        "Downloaded blobs for block: {}",
+        block.message.body.execution_payload.block_number
+    );
+
+    let state = get_state(&checkpoint_sync_url, slot).await?;
+    info!(
+        "Downloaded State with root: {}. Slot: {}",
+        state.state_root(),
+        slot
+    );
+    ensure!(block.message.slot == state.slot, "Slot mismatch");
+
+    ensure!(block.message.state_root == state.state_root());
+    let mut store = get_forkchoice_store(state, block.message, db)?;
+
+    let time = network_spec().genesis.genesis_time + SECONDS_PER_SLOT * (slot + 1);
+    on_tick(&mut store, time)?;
+    info!("Initial Sync complete");
+
+    Ok(())
+}
+
+/// Fetch initial state from trusted RPC
+pub async fn get_state(rpc: &Url, slot: u64) -> anyhow::Result<BeaconState> {
+    let client = reqwest::Client::new();
+    let state = client
+        .get(format!("{rpc}eth/v2/debug/beacon/states/{slot}"))
+        .header(ACCEPT, HeaderValue::from_static("application/octet-stream"))
+        .send()
+        .await?
+        .bytes()
+        .await?;
+
+    BeaconState::from_ssz_bytes(&state)
+        .map_err(|err| anyhow!("Unable to decode state from ssz bytes: {err:?}"))
+}
+
+/// Fetch initial block from trusted RPC
+pub async fn fetch_finalized_block(rpc: &Url) -> anyhow::Result<SignedBeaconBlock> {
+    let client = reqwest::Client::new();
+    let raw_bytes = client
+        .get(format!("{rpc}eth/v2/beacon/blocks/finalized"))
+        .header(ACCEPT, HeaderValue::from_static("application/octet-stream"))
+        .send()
+        .await?
+        .bytes()
+        .await?;
+
+    SignedBeaconBlock::from_ssz_bytes(&raw_bytes)
+        .map_err(|err| anyhow!("Unable to deocde block from ssz byttes: {err:?}"))
```

Comment:
```suggestion
        .map_err(|err| anyhow!("Unable to decode block from ssz bytes: {err:?}"))
```

small spelling issues